### PR TITLE
Subscription only registers active

### PR DESCRIPTION
**Pull request by fil512**
Fixed a bug in standalone subscription subscriber: It was adding REQUESTED subscriptions to the active subscription registry. (Only ACTIVE subscriptions should be added.)

### CODE REVIEW

**Comment by fil512** [comment]
@jamesagnew ready for your review

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/retry/Retrier.java [hunk 1]
```diff
@@ -20,44 +20,41 @@
...
-		mySecondsBetweenRetries = theSecondsBetweenRetries;
-		myDescription = theDescription;
+
+		myRetryTemplate = new RetryTemplate();
+
+		ExponentialBackOffPolicy backOff = new ExponentialBackOffPolicy();
+		backOff.setInitialInterval(500);
+		backOff.setMaxInterval(DateUtils.MILLIS_PER_MINUTE);
+		backOff.setMultiplier(2);
+		myRetryTemplate.setBackOffPolicy(backOff);
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/retry/Retrier.java]: Man... Spring is just so nice for this kind of thing.

File: hapi-fhir-jpaserver-subscription/pom.xml [hunk 2]
```diff
@@ -97,6 +97,11 @@
 			<artifactId>jetty-servlet</artifactId>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-test</artifactId>
+			<scope>test</scope>
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-subscription/pom.xml]: Out of curiosity, why is this needed?

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/ResourceModifiedMessage.java [hunk 3]
```diff
@@ -166,4 +166,12 @@ private static boolean payloadContainsNoPlaceholderReferences(FhirContext theCtx
 		return true;
 	}
 
+	@Override
+	public String toString() {
+			String resourceId = myPayloadId;
+			if (resourceId == null) {
+				resourceId = myId;
+			}
+			return "ResourceModified Message { " + myOperationType + ", " + resourceId + "}";
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/ResourceModifiedMessage.java]: FWIW, we typically do these using ToStringBuilder. Intellij will actually do all the work for you if you want. 

[Here's a random example](https://github.com/jamesagnew/hapi-fhir/blob/master/hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/entity/ResourceReindexJobEntity.java#L127)

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/cache/SubscriptionCanonicalizer.java [hunk 4]
```diff
@@ -292,4 +292,10 @@ public void setMatchingStrategyTag(FhirContext theFhirContext, IBaseResource the
 		}
 		meta.addTag().setSystem(SubscriptionConstants.EXT_SUBSCRIPTION_MATCHING_STRATEGY).setCode(value).setDisplay(display);
 	}
+
+	public String getSubscriptionStatus(IBaseResource theSubscription) {
+		final IPrimitiveType<?> status = myFhirContext.newTerser().getSingleValueOrNull(theSubscription, SubscriptionConstants.SUBSCRIPTION_STATUS, IPrimitiveType.class);
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/cache/SubscriptionCanonicalizer.java]: Careful of NPE here

Comment (jamesagnew) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/cache/SubscriptionCanonicalizer.java]: Actually, the NPE would be on the line below...

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/subscriber/ResourceDeliveryJsonMessage.java [hunk 5]
```diff
@@ -54,4 +54,8 @@ public void setPayload(ResourceDeliveryMessage thePayload) {
 		myPayload = thePayload;
 	}
 
+	@Override
+	public String toString() {
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/subscriber/ResourceDeliveryJsonMessage.java]: Recommend to use ToStringBuilder

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/subscriber/ResourceDeliveryMessage.java [hunk 6]
```diff
@@ -114,4 +114,10 @@ public void setPayloadId(IIdType thePayloadId) {
 			myPayloadId = thePayloadId.getValue();
 		}
 	}
+
+	@Override
+	public String toString() {
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/subscriber/ResourceDeliveryMessage.java]: Recommend to use ToStringBuilder

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/subscriber/ResourceModifiedJsonMessage.java [hunk 7]
```diff
@@ -55,4 +55,8 @@ public void setPayload(ResourceModifiedMessage thePayload) {
 		myPayload = thePayload;
 	}
 
+	@Override
+	public String toString() {
```
Comment (jamesagnew) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/subscriber/ResourceModifiedJsonMessage.java]: Recommend to use ToStringBuilder

File: hapi-fhir-jpaserver-subscription/pom.xml (hunk 2 above)
Comment (fil512) [reply hapi-fhir-jpaserver-subscription/pom.xml]: For @MockBean in my test

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/subscriber/ResourceDeliveryJsonMessage.java (hunk 5 above)
Comment (fil512) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/subscriber/ResourceDeliveryJsonMessage.java]: Cool.  I will change these

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/ResourceModifiedMessage.java (hunk 3 above)
Comment (fil512) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/ResourceModifiedMessage.java]: done

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/cache/SubscriptionCanonicalizer.java (hunk 4 above)
Comment (fil512) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/module/cache/SubscriptionCanonicalizer.java]: done
